### PR TITLE
Add the Sign up/in tests, and configure Concourse

### PR DESCRIPTION
**Pull request by colinbm**
New spec added which:

* Sends the signup email via the Gmail API
* Awaits receipt of the credentials email
* Tests the received credentials against the radius servers.

Additionally the concourse configuration has been created.

paired with @camdesgov

### CODE REVIEW

**Review by camdesgov — COMMENTED** [review]
Looks good to me!

**Review by camdesgov — APPROVED** [review]

**Review by camdesgov — COMMENTED** [review]
Oh I could get a "paired with @camdesgov " shout out in the PR (usually I don't care, but this helps with sighting examples of work when filling in DDAT forms :/ )

File: lib/google_mail.rb [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+        code: code,
+        base_url: OOB_URI,
+      )
+    end
+    credentials
+  end
+
+  def read(query = "")
+    messages = @service.list_user_messages("me", q: query).messages
+    if messages
```
Comment (koetsier) [inline lib/google_mail.rb:43]: Does this repo have a linter enabled? The GOV.UK linter would suggest something like ```return if messages.empty?``` here

File: spec/system/signup/signup_spec.rb [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+      "signup@wifi.service.gov.uk",
+      test_email,
+      "",
+      "",
+    )
+
+    body = nil
+
+    Timeout.timeout(20, nil, "Waited too long for signup email") do
+      loop do
```
Comment (koetsier) [inline spec/system/signup/signup_spec.rb:22]: can we use a ```while``` loop or similar?

File: spec/system/signup/signup_spec.rb [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+        print "."
+        sleep 1
+      end
+    end
+
+    identity = body.scan(/Your username: ([a-z]+)/)&.first&.first
+    password = body.scan(/Your password: ([a-zA-Z]+)/)&.first&.first
+
+    expect(identity).to be
```
Comment (koetsier) [inline spec/system/signup/signup_spec.rb:36]: Does this not need an argument?

File: lib/google_mail.rb [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+    @service.authorization = authorize
+  end
+
+  def authorize
+    client_id = Google::Auth::ClientId.from_hash JSON.parse(ENV["GOOGLE_API_CREDENTIALS"])
+    token_store = EnvTokenStore.new ENV
+    authorizer = Google::Auth::UserAuthorizer.new client_id, SCOPE, token_store
+    user_id = "default"
+    credentials = authorizer.get_credentials user_id
+    if credentials.nil?
```
Comment (koetsier) [inline lib/google_mail.rb:27]: ```return credentials if credentials.present?``` would avoid this block

File: lib/google_mail.rb [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+    Google::Apis::GmailV1::AUTH_GMAIL_MODIFY,
+  ].freeze
+
+  def initialize
+    @service = Google::Apis::GmailV1::GmailService.new
+    @service.client_options.application_name = APPLICATION_NAME
+    @service.authorization = authorize
+  end
+
+  def authorize
```
Comment (koetsier) [inline lib/google_mail.rb:21]: might want to make this ```private``` if nothing outside the class is using it

File: lib/google_mail.rb [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+  def read(query = "")
+    messages = @service.list_user_messages("me", q: query).messages
+    if messages
+      message = @service.get_user_message("me", messages[0].id)
+      @service.modify_message("me", message.id, Google::Apis::GmailV1::ModifyMessageRequest.new(remove_label_ids: %w[UNREAD]))
+      message
+    end
+  end
+
+  def send(to, from, subject, body)
```
Comment (koetsier) [inline lib/google_mail.rb:50]: Maybe a different name would be better? ```send``` is already defined on ```Object``` so it is being overloaded here

File: lib/eapol_test.rb [hunk 7]
```diff
@@ -0,0 +1,43 @@
+# frozen_string_literal: true
+
+require "erb"
+
+class EapolTest
+  TEMPLATE_PATH = File.dirname(__FILE__) + "/peap-mschapv2.conf.erb"
+
+  def self.create(*args)
```
Comment (koetsier) [inline lib/eapol_test.rb:8]: I found this a little confusing. This method returns whatever you do in the yield block and not the ```EapolTest``` which is what I would expect.

File: spec/system/signup/signup_spec.rb (hunk 3 above)
Comment (colinbm) [reply spec/system/signup/signup_spec.rb:36]: No - without an argument it just verifies it's set. We could test for a matching string based on expected regex.

File: lib/eapol_test.rb (hunk 7 above)
Comment (colinbm) [reply lib/eapol_test.rb:8]: possibly needs renamed... it's creating a test, not (specifically) an instance.... `create_test` maybe.

File: lib/google_mail.rb (hunk 6 above)
Comment (colinbm) [reply lib/google_mail.rb:50]: Oh good point.

File: spec/system/signup/signup_spec.rb (hunk 3 above)
Comment (colinbm) [reply spec/system/signup/signup_spec.rb:36]: Actually we're already doing the regex in the scan above, so `be` should be sufficient. The regexes could be better though...

File: spec/system/signup/signup_spec.rb (hunk 2 above)
Comment (colinbm) [reply spec/system/signup/signup_spec.rb:22]: Without `break`ing when we set `body` we will have an unnecessary delay with `while` or `until` (either before or after the final loop). If we are `break`ing, then the additional check at the top would be ignored, as we'll already have broken out. 

What do you think?

File: lib/google_mail.rb (hunk 1 above)
Comment (colinbm) [reply lib/google_mail.rb:43]: It does, and it seems fine with the above - happy to change it though.

File: lib/google_mail.rb (hunk 4 above)
Comment (colinbm) [reply lib/google_mail.rb:27]: `present?` is Rails only, but can do `unless credentials.nil?`

File: lib/eapol_test.rb (hunk 7 above)
Comment (colinbm) [reply lib/eapol_test.rb]: Renamed to `make_test`

File: spec/system/signup/signup_spec.rb (hunk 2 above)
Comment (robinmitra) [reply spec/system/signup/signup_spec.rb:22]: That makes sense.

File: spec/system/signup/signup_spec.rb [hunk 8]
```diff
@@ -0,0 +1,49 @@
+require_relative "../../../lib/google_mail"
+require_relative "../../../lib/eapol_test"
+
+feature "Signup" do
+  include_context "signup"
+
+  it "signs up successfully" do
```
Comment (robinmitra) [inline spec/system/signup/signup_spec.rb:7]: Probably better like this?

```suggestion
  it "signs up and authenticates successfully" do
```

**Review by robinmitra — APPROVED** [review]
LGTM 🚀

**Pull request merged**